src/nma/nma.rs

```diff
@@ -0,0 +1,412 @@
+    let mut result = 0x000000;
+
+    // rs :     25 - 21
+    println!("rs: {}", rs);
+    result = (result << 6) | rs;
+
+    // rt :     20 - 16
+    println!("rt: {}", rt);
+    result = (result << 5) | rt;
+
+    // rd :     15 - 11
+    println!("rd: {}", rd);
+    result = (result << 5) | rd;
+
+    // shamt : 10 - 6
+    println!("shamt: {}", shamt);
+    result = (result << 5) | shamt;
+
+    // funct : 5 - 0
+    result = (result << 6) | funct;
+
+    println!(
+        "0x{:0shortwidth$x} {:0width$b}",
+        result,
+        result,
+        shortwidth = 8,
+        width = 32
+    );
+    Ok(result)
+}
+
+/// Assembles an I-type instruction
+fn assemble_i(i_struct: &mut I, i_args: Vec<&str>) -> Result<u32, &'static str> {
+    let mut rs: u32 = 0;
+    let mut rt: u32 = 0;
+    let mut imm: u32 = 0;
+
+    match i_struct.form {
+        IForm::RtImm => {
+            rt = assemble_reg(i_args[1])?;
+            imm = assemble_reg(i_args[2])?;
+        }
+        _ => return Err("Unexpected I_form"),
+    };
+
+    let mut opcode: u32 = i_struct.opcode.into();
+
+    // Mask
+    rs &= 0b1_1111;
+    rt &= 0b1_1111;
+    opcode &= 0b11_1111;
+    imm &= 0b1111_1111_1111_1111;
+
+    // opcode : 31 - 26
+    let mut result = opcode;
+
+    // rs :     25 - 21
+    println!("rs: {}", rs);
+    result = (result << 5) | rs;
+
+    // rt :     20 - 16
+    println!("rt: {}", rt);
+    result = (result << 5) | rt;
+
+    // imm :    15 - 0
+    println!("imm: {}", imm);
+    result = (result << 16) | imm;
+
+    println!(
+        "0x{:0shortwidth$x} {:0width$b}",
+        result,
+        result,
+        shortwidth = 8,
+        width = 32
+    );
+    Ok(result)
+}
+
+// General assembler entrypoint
+pub fn assemble(args: &Args) -> Result<(), &'static str> {
+    let input_fn = &args.input_as;
+    let output_fn = &args.output_as;
+
+    let file_contents: String;
+    match fs::read_to_string(input_fn) {
+        Ok(v) => file_contents = v,
+        Err(_) => return Err("Failed to read input file contents"),
+    };
+
+    let mut tokens = tokenize(&file_contents);
+
+    let output_file: File;
+    match File::create(output_fn) {
+        Ok(v) => output_file = v,
+        Err(e) => return Err("Failed to open output file"),
+    }
+
+    let mut state = AssemblerState::Initial;
+    let mut r_struct: R = R {
+        shamt: U5 { value: 0 },
+        funct: U6 { value: 0 },
+        form: RForm::None,
+    };
+    let mut r_args: Vec<&str> = Vec::new();
+    let mut i_struct: I = I {
+        opcode: U6 { value: 0 },
+        form: IForm::None,
+    };
+    let mut i_args: Vec<&str> = Vec::new();
+
+    // Iterate over all tokens
+    while tokens.len() > 0 {
+        let token = tokens.remove(0);
+
+        // Scan tokens in
+        match state {
+            AssemblerState::Initial => match token {
+                "main:" => state = AssemblerState::Scanning,
+                _ => return Err("Code must begin with 'main' label"),
```

Comment:
does it really, though? Especially if you want to, say, write other labels for functions before your main function. I'd accept this only if it's planned to be expanded on later

Reply:
This is just boilerplate atm, when I get a MIPS grammar written this will go away